Start and wait for a checkup Job to finish

ormergi opened the pull request:
This PR introduce the logic for actually executing a checkup in a cluster.

With this PR changes the framework will create a batch Job to run the checkup image
and waits for it to finish.

A Job is considered as finished when its status has 'Completed' or 'Failed' condition.

orelmisan commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -225,6 +227,39 @@ func TestCheckupTeardownShould(t *testing.T) {
+
+func TestCheckupRunShouldCreateJob(t *testing.T) {
+	checkupRunTestCases := []checkupRunTestCase{
+		{description: "with no addition env vars"},
```
*additional

ormergi replied:
Done

orelmisan commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -225,6 +227,39 @@ func TestCheckupTeardownShould(t *testing.T) {
+	envVars     []corev1.EnvVar
+}
+
+func TestCheckupRunShouldCreateJob(t *testing.T) {
```
nit: Please consider changing this function's name to `TestCheckupRunShouldCreateAJob`

orelmisan commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -225,6 +227,39 @@ func TestCheckupTeardownShould(t *testing.T) {
+	}
+}
+
+func TestCheckupRunShouldFailWhen(t *testing.T) {
```
Is it possible to simulate the following errors using the fake client?
1. Failure to fetch the image.
2. Checkup exiting with a non zero exit code.

Do we want to simulate these kinds of errors in the unit tests?

orelmisan replied:
Can we differentiate between these problems?

EdDev replied:
@orelmisan , why do we need to differentiate between different errors? Is there a different action to take if the error differs?
I would even claim that you do not even covered here all the error types that can be raised, moreover, there may be others added (or removed) in the future.
As I see it, if there is no differentiation on the production code about what to do when a specific error is detected, there is no point in covering it in tests.

orelmisan replied:
Can we deduce it from another source?

This is important for loading (or not) the checkup's results from its ConfigMap.

orelmisan reviewed (commented):
Thank you for the PR @ormergi.
Please see the inline comments for the first commit.

orelmisan commented on kiagnose/internal/checkup/checkup.go:
```diff
@@ -285,6 +285,10 @@ func (c *Checkup) Run() error {
 		return err
 	}
 
+	if c.job, err = job.WaitForJobToFinish(c.client.BatchV1(), c.job, c.jobTimeout); err != nil {
+		return err
+	}
+
 	return nil
```
I guess that this will be the integration point with the component that will read the checkup results from the results ConfigMap.
We need to think on which errors can we try to read the results.

EdDev replied:
I think that the question was about how this "waiting" here integrates with the need to record its run status on the result ConfigMap.

So let me ask this in a more specific manner:
- If the wait ends without a timeout, is the job still around to query its status?

orelmisan replied:
This could lead to reading an empty ConfigMap object.

ormergi replied:
Sorry for the confusion this is correct ttlSecondsAfterFinished will trigger the job to delete and activeDeadlineSeconds trigger it to stop.

orelmisan commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -225,6 +227,39 @@ func TestCheckupTeardownShould(t *testing.T) {
+			testClient := newNormalizedFakeClientset()
+			testCheckup := checkup.New(testClient, &config.Config{Image: testImage, Timeout: testTimeout, EnvVars: testCase.envVars})
+
+			assert.NoError(t, testCheckup.Run())
```
How is the checkup running successfully without the `Setup()`?

orelmisan replied:
I just don't understand how the checkup is run without its supporting objects created by Setup().
Can you please elaborate?

orelmisan commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -225,6 +227,39 @@ func TestCheckupTeardownShould(t *testing.T) {
+		testClient.injectCreateErrorForResource(jobResource, expectedErr)
+		testCheckup := checkup.New(testClient, &config.Config{Image: testImage})
+
+		assert.ErrorContains(t, testCheckup.Run(), expectedErr)
```
How is the checkup running without calling setup()?

orelmisan commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -36,3 +41,54 @@ func Create(client batchv1client.BatchV1Interface, job *batchv1.Job) (*batchv1.J
+	const JobNameLabel = "job-name"
+
+	jobLabel := fmt.Sprintf("%s=%s", JobNameLabel, job.Name)
+	jobWatcher, err := client.Jobs(job.Namespace).Watch(context.Background(), metav1.ListOptions{LabelSelector: jobLabel})
```
Can you please consider using an informer?
It should have an advantage over raw watch, for example it can recover from a momentary communication failure.

ormergi replied:
As we concluded offline due to the complexity using an informer brings using bare watcher is good enough for now.
I posted new PR for using informer https://github.com/kiagnose/kiagnose/pull/32

orelmisan replied:
Thank you for the effort.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/f7331e95b4d7953b82ffc44cea0ed81a823eadc4..dba2c4e0f011978ffff344c657f8efd4e2269ce3): Addressed code review comments.
Also use context with timeout when waiting for the checkup job to finish.

EdDev commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -278,6 +312,16 @@ func newTestEnvVars() []corev1.EnvVar {
+			Name:      name,
+			Namespace: namespace},
+		Status: batchv1.JobStatus{
+			Conditions: []batchv1.JobCondition{
+				{Type: condition}}}}
```
nit: Can you place this in one line?
It is hard to see the relation between opened and closed braces.. It will be nice to see blocks like this:
```
foo{
    boo{
        zoo{..{}}
    }
}
```

I wander if we can enforce this with a formatter.

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -36,3 +40,55 @@ func Create(client batchv1client.BatchV1Interface, job *batchv1.Job) (*batchv1.J
 }
+
+func WaitForJobToFinish(client batchv1client.BatchV1Interface, job *batchv1.Job, timeout time.Duration) (*batchv1.Job, error) {
+	const JobNameLabel = "job-name"
```
Where is this coming from?

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -36,3 +40,55 @@ func Create(client batchv1client.BatchV1Interface, job *batchv1.Job) (*batchv1.J
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("'%s/%s' Job watcher obtained", job.Namespace, job.Name)
```
Why do we need this? If it was not obtained, we get an error, no?

ormergi replied:
Done

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -36,3 +40,55 @@ func Create(client batchv1client.BatchV1Interface, job *batchv1.Job) (*batchv1.J
+			if !ok {
+				continue
+			}
+			if jobStatus, err := json.MarshalIndent(job.Status, "", " "); err == nil {
+				log.Printf("received job event '%s/%s': \n%v\n", job.Namespace, job.Name, string(jobStatus))
+			}
```
If we will log this for every job change, it will be too verbose IMO.
I do not think this needs to exist in the production code.

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -36,3 +40,55 @@ func Create(client batchv1client.BatchV1Interface, job *batchv1.Job) (*batchv1.J
+
+func finished(job *batchv1.Job) bool {
+	for _, condition := range job.Status.Conditions {
+		if condition.Type == batchv1.JobComplete || condition.Type == batchv1.JobFailed {
+			return true
+		}
```
`return condition.Type == batchv1.JobComplete || condition.Type == batchv1.JobFailed`

ormergi replied:
Note that Job status conditions can be `Completed` `Failed` or `Suspended`, other than that it seems that its possible to add any other arbitrary condition to the Job status (according to doc status.conditions can be patched)
https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/job-v1/#JobStatus

EdDev replied:
Based on that information, this is not complete then. Right?

ormergi replied:
From what I tested a Job is considered to be finished when `Complete` or `Failed` conditions exists,
this is why we look for these conditions at Job status.

EdDev replied:
I think it says you need to check the `conditions.status` value for `true` as well.

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -36,3 +40,55 @@ func Create(client batchv1client.BatchV1Interface, job *batchv1.Job) (*batchv1.J
+	defer watcher.Stop()
+	for {
+		select {
+		case event := <-eventsCh:
```
`watcher.Stop()` closes the channel?

ormergi reviewed (commented):
@orelmisan @EdDev please take another look I resolved some code review comments and answered questions.

EdDev commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -389,6 +433,22 @@ func assertClusterRoleBindingsCreated(t *testing.T, testClient testsClient, clus
 
+func assertJobCreated(t *testing.T, testClient *testsClient, image string, timeout time.Duration, envVars []corev1.EnvVar) {
+	gvr := schema.GroupVersionResource{Group: batchv1.GroupName, Version: "v1", Resource: jobResource}
+	actualJob, err := testClient.Tracker().Get(gvr, checkup.NamespaceName, checkup.JobName)
```
Is it possible to use a simple `Get` here? Do you need the `Tracker()` in this check?

EdDev commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -389,6 +433,22 @@ func assertClusterRoleBindingsCreated(t *testing.T, testClient testsClient, clus
 	assert.Subset(t, actualClusterRoleBindings, expectedClusterRoleBindings)
 }
 
+func assertJobCreated(t *testing.T, testClient *testsClient, image string, timeout time.Duration, envVars []corev1.EnvVar) {
```
This helper does 3 things, hiding a lot of things from the tests, which in this case could have assisted the reader to understand better.

WDYT about this (at the test):
```
createdJob := ...
expectedJob := ...
assert.Equal(t, createdJob, expectedJob)
```

EdDev commented on kiagnose/internal/checkup/checkup_test.go:
```diff
@@ -267,6 +295,12 @@ func TestCheckupRunShouldFailWhen(t *testing.T) {
+	t.Run("Job wont finish on time", func(t *testing.T) {
+		testCheckup := checkup.New(fake.NewSimpleClientset(), &config.Config{Image: testImage, Timeout: time.Nanosecond})
+
+		assert.ErrorContains(t, testCheckup.Run(), context.DeadlineExceeded.Error())
```
It will be better to assert on the error type/indentity if possible. Is it possible without changing the production code?

EdDev replied:
Lets proceed as is, this is good enough.

ormergi reviewed (commented):
Addressed code review comments

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/ddd9397e459fcae68131837cbc7de147393b015a..2c5d72ed1160fc06d54ad2820f72cb4778a40eee): Check conditions.status when watching the checkup Job.

EdDev commented on kiagnose/internal/checkup/job/job.go:
```diff
@@ -0,0 +1,95 @@
+
+func finished(job *batchv1.Job) bool {
+	for _, condition := range job.Status.Conditions {
+		if condition.Type == batchv1.JobComplete || condition.Type == batchv1.JobFailed &&
+			condition.Status == corev1.ConditionTrue {
+			return true
+		}
```
This seems incorrect to me: I think the `||` logical expression needs to be parenthesized.

Please check how you can cover this bug in the tests.

ormergi commented:
[Changes](https://github.com/kiagnose/kiagnose/compare/013bd721262d165867fc2c756baf6be0b1a2ab61..65acadcccf133d5b606016f10f9056f006367250): Change the injected Job event condition definition to be explicit on tests, and cover the case that was mentioned at https://github.com/kiagnose/kiagnose/pull/27#discussion_r858336398

EdDev reviewed (approved):
Thank you!
